### ISSUE TITLE
Slowness at each keystroke

### ISSUE THREAD

**Issue by djes** [qgis/QGIS#50905]
### What is the bug or the crash?

Tested on QGIS 3.16-3.28, Windows 10. Asked on qgis-user-mailing list and some users experience the same thing.

There are two problems : 

- When typing in the print layout, for example to change a legend entry name, if the map is visible in the background, each keystroke is slow. A hint is to zoom in or hide the map ;
- When typing on a big project with relations in a form text field with multiple tabs, each keystroke is a pain, like 1-2s to react.

The feeling is that something happens at each keystroke, as if there was an event launched, like a full refresh. This is not linked to calculated formula (tried without).

Wouldn't it be possible to stop refresh at each keystroke when typing in a field, but only at the end, when pressing return or leaving the field ?

I precise that I have a good computer with a lot of RAM.

### Steps to reproduce the issue

Only big projects are concerned, on Windows, with DB and several layers and relations, so it's hard to give a full working example. Anyway if someone explains me how to debug events, I can try to put a breakpoint somewhere to help.

### Versions

3.16
3.22.12-Białowieża
3.28.0-Firenze

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by ReghardNell** [comment]
I have noticed the same.

Also running QGIS 3.16-3.28, Windows 10.

Resorted to typing in notepad and pasting labels to minimize delays as mentioned above.

Réghard

**Comment by andreasneumann** [comment]
@djes - can you please open different issues for each problem you encounter? Print layouts and forms are totally different issues. They probably have nothing in common.

Also: please always provide a sample project configuration and data. General reports like this won't be looked at by developers, without a project and data to reproduce.

**Comment by baswein** [comment]
I am also encountering editing legend text to be painfully slow (15 sec per letter) on both windows and mac machines. But it our case the map being visible doesn't seem to make a difference. If the legend itself is off when the text is edited then it doesn't slow down but then take a minute or two to show again when turned back on. I'll try to boil down the issue and see if I can get a sample dataset.

**Comment by nyalldawson** [comment]
> I'll make a video screen capture with my smartphone.

We definitely need a reproducible project to solve this. It's not a widespread issue and a screencast alone won't be enough.

**Comment by djes** [comment]
> > I'll make a video screen capture with my smartphone.
> 
> We definitely need a reproducible project to solve this. It's not a widespread issue and a screencast alone won't be enough.

I have attached a working local example.

**Comment by baswein** [comment]
I can confirm that the qgis_layout_bug test file is also slow on my system. About 2 seconds a letter.

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>


QGIS version | 3.28.0-Firenze | QGIS code revision | ed3ad0430f3
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.9.5
GDAL/OGR version | 3.5.2
PROJ version | 9.1.0
EPSG Registry database version | v10.074 (2022-08-01)
GEOS version | 3.10.3-CAPI-1.16.1
SQLite version | 3.39.4
PDAL version | 2.4.3
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.6
QScintilla2 version | 2.13.1
OS version | Windows 10 Version 2009
  |   |   |  
Active Python plugins
db_manager | 0.1.20
grassprovider | 2.12.99
MetaSearch | 0.3.6
otbprovider | 2.12.99
processing | 2.12.99
sagaprovider | 2.12.99

</body></html>QGIS version
3.28.0-Firenze
QGIS code revision
[ed3ad0430f3](https://github.com/qgis/QGIS/commit/ed3ad0430f3)
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.5.2
PROJ version
9.1.0
EPSG Registry database version
v10.074 (2022-08-01)
GEOS version
3.10.3-CAPI-1.16.1
SQLite version
3.39.4
PDAL version
2.4.3
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009




Active Python plugins
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
otbprovider
2.12.99
processing
2.12.99
sagaprovider
2.12.99

**Comment by nyalldawson** [comment]
Seems specific to Windows -- I can't reproduce on linux

**Comment by baswein** [comment]
My colleague is also experiencing it on a mac. I'll have her try it with the test file tomorrow.  

Also some other users experiencing the issue.
https://gis.stackexchange.com/questions/396448/editing-legend-text-very-demanding
https://www.reddit.com/r/QGIS/comments/xd4fct/extremely_slow_chug_with_legend_updates_in_layout/

**Comment by baswein** [comment]
She tried the test file on her MAC and doesn't have a problem with it. But has had problems with other files. If we can isolate the problem with other files I'll make a test file for MAC's.

**Comment by djes** [comment]
I've added the fact that any field in the attribute table is concerned by the slowness. I also noted that in 3.29, a calculated field is updated only when changing value, and that it fires a whole screen/fields refresh at each keystroke.

I think it's an old low level behavior and that it should be changed, because if it would have an interest with an ideal configuration, it's a pain with a slower real world project...

**Comment by djes** [comment]
I recently noted that I have slowness for the north arrow. I have a SVG library located on network drives, not used for the north arrow, but the simple fact to add the north arrow on a template is awfully slow. Is there a full library refresh at this moment ? Removing the network drive from the list of SVG folders resolves the problem.

**Comment by baswein** [comment]
@djes Does removing the SVG network share resolve the slow legend text issues as well?

**Comment by djes** [comment]
@baswein no, legend and attribute table are slow as usual. The more on screen, the more it is slow. 

I think that they keystroke event generates a full screen refresh and that the SVG moving includes some SVG panel refresh event...

**Comment by djes** [comment]
I've noticed that our legend contains hashes based on SVG. Removing them has fastened things a lot in print layout. I'll investigate more on this side. Of course it changes nothing on attributes slowness.

**Comment by luca76** [comment]
> I've noticed that our legend contains hashes based on SVG. Removing them has fastened things a lot in print layout. I'll investigate more on this side. Of course it changes nothing on attributes slowness.

Can you explain me what are hashes based on SVG? And how did you remove them?

**Comment by djes** [comment]
It's a nicely designed repetitive vector pattern done with Inkscape and
used to fill polygon.  I've just removed these SVG patterns from the fill
settings.
It seems to me that QGIS try to redraw everything at each event, even a
simple keystroke in a text field. It waits for the SVG drawing to be
finished before handling any other user event.
User event should be priorised and should stop drawing. Surely the SVG
drawing has no breaking signal.

Regards,

Jésahel

**Comment by djes** [comment]
> @djes may I ask you to retest with the latest versions? #51681 should improve situation.

With the given project, still slow, as this heavy legend text is still fully updated at each keystroke in WYSIWYG. I have not tested with SVG or heavy symbols.
